aide-de-camp/src/runner/job_runner.rs

```diff
@@ -50,26 +84,77 @@ impl<Q> JobRunner<Q>
 where
     Q: Queue + 'static,
 {
-    /// This function runs forever unless built-in Semaphore gets closed. Which never happens.
-    /// Later there might be a way to stop it.
+    /// Run the job runner with the desired polling interval. This method blocks forever.
     pub async fn run(&mut self, interval: chrono::Duration) -> Result<(), QueueError> {
+        // Pass in a future that never completes to the shutdown handler so the server can run until program termination
+        self.run_with_shutdown(interval, future::pending()).await
+    }
+
+    /// Run the job runner with the desired polling interval.
+    /// This method blocks until the provided shutdown future completes.
+    pub async fn run_with_shutdown<F: Future<Output = ()>>(
+        &mut self,
+        interval: chrono::Duration,
+        shutdown: F,
+    ) -> Result<(), QueueError> {
+        let worker_handles = FuturesUnordered::new();
+        let cancellation_token = CancellationToken::new();
+        let mut shutdown = Box::pin(shutdown);
         loop {
             let semaphore = self.semaphore.clone();
-            let permit = semaphore
-                .acquire_owned()
-                .await
-                .context("Semaphore closed while running")?;
-            let queue = self.queue.clone();
-            let processor = self.processor.clone();
-            tokio::spawn(async move {
-                let _permit = permit;
-                let queue = queue;
```

Comment:
I removed these two lines:
```rust
let queue = queue;
let processor = processor;
```
from here when I was moving this over because I don't think they're necessary, but let me know if I missed something and I can add this back.